File: src/renderer/src/state/workspaceStatus.ts
import { create } from 'zustand'

// Per-terminal agent/activity state, keyed so we can roll it up per workspace for
// the workspace rail cards. Fed by TerminalPanel from the pty status/bell/done
// signals. (A richer descriptive status — "Claude is waiting for your input" —
// will arrive with the OSC agent-status feature; this is the activity rollup.)
export type PaneActivity = 'idle' | 'busy' | 'attn'

interface PaneStatus {
  workspace: string
  busy: boolean
  attention: boolean
}

interface WorkspaceStatusState {
  panes: Record<string, PaneStatus>
  setPane: (workspace: string, paneId: number, patch: Partial<Omit<PaneStatus, 'workspace'>>) => void
  clearPane: (workspace: string, paneId: number) => void
}

export const useWorkspaceStatus = create<WorkspaceStatusState>((set) => ({
  panes: {},
  setPane: (workspace, paneId, patch) =>
    set((st) => {
      const key = `${workspace}|${paneId}`
      const prev = st.panes[key] ?? { workspace, busy: false, attention: false }
      return { panes: { ...st.panes, [key]: { ...prev, workspace, ...patch } } }
    }),
  clearPane: (workspace, paneId) =>
    set((st) => {
      const key = `${workspace}|${paneId}`
      if (!(key in st.panes)) return st
      const next = { ...st.panes }
      delete next[key]
      return { panes: next }
    })
}))

// Roll the per-pane state up to one activity level for a workspace.
export function rollupActivity(
  panes: Record<string, PaneStatus>,
  workspace: string
): PaneActivity {
  let busy = false
  for (const p of Object.values(panes)) {
    if (p.workspace !== workspace) continue
    if (p.attention) return 'attn'
    if (p.busy) busy = true
  }
  return busy ? 'busy' : 'idle'
}

File: Sources/Riven/Workspace/WorkspaceStatus.swift
import Foundation

// Per-workspace activity rollup (riven's state/workspaceStatus.ts). Each terminal
// pane reports busy / needs-attention; a workspace rolls up to the most urgent:
// needs-you > busy > idle. 상태 어휘 자체는 [[AgentStatus]] 하나로 통일돼 있다
// (UI/AgentStatus.swift) — 여기서는 워크스페이스 단위로 접기만 한다.

final class WorkspaceStatus {
    static let shared = WorkspaceStatus()
    private init() {}

    private struct Pane { let ws: String; var busy: Bool; var attn: Bool }
    private var panes: [String: Pane] = [:]     // key = "ws|paneId"
    var onChange: ((_ workspace: String) -> Void)?

    func setPane(ws: String, pane: String, busy: Bool? = nil, attn: Bool? = nil) {
        let key = "\(ws)|\(pane)"
        var cur = panes[key] ?? Pane(ws: ws, busy: false, attn: false)
        if let busy { cur.busy = busy }
        if let attn { cur.attn = attn }
        panes[key] = cur
        onChange?(ws)
    }
    func clearPane(ws: String, pane: String) { panes["\(ws)|\(pane)"] = nil; onChange?(ws) }

    func rollup(_ ws: String) -> AgentStatus {
        let ps = panes.values.filter { $0.ws == ws }
        if ps.contains(where: { $0.attn }) { return .done }
        if ps.contains(where: { $0.busy }) { return .busy }
        return .idle
    }
}
